### ISSUE TITLE
Clarification of GPT results in Table 2 on MIMIC-III data

### ISSUE THREAD

**Issue by jameszhou-gl** [btheodorou99/HALO_Inpatient#21]
Hi authors,

Thanks for sharing the code. I'm curious about the GPT results in Table 2, i.e., on MIMIC-III data, GPT-2 model achieves 1.26 on `record length mean` and 9.67 on `visit length mean`. 
However, when I used exactly the same [baselines/gpt](https://github.com/btheodorou99/HALO_Inpatient/blob/main/baselines/gpt/train_gpt.py) code and generated the synthetic data, I found the synthetic data primarily consists of only one visit per record with hundreds of diagnosis codes for each visit. In this case, `Record length mean` is 1 and `Visit length mean` is 325. 

Have you ever met a similar issue? Seems like the gpt model might not learn the diagnosis code sequence effectively? It just generates the code from the vocabulary but `END_OF_VISIT` token appears very infrequently. This leads to unnaturally long visit lengths compared to the training data.

 Thank you so much if any assistance available.

**Comment by btheodorou99** [comment]
Hello, we did have some challenges while running the experiments where our special tokens were being either improperly written or improperly masked. We fixed these issues on our end, but there is a chance that refactoring the code for publishing may have reintroduced some of them. Specifically, make sure that the code is properly using the config from the gpt folder as opposed to the main folder and that the numbers for code_vocab_size and total_vocab_size correctly match your dataset (we have had some other people use a slightly different version of MIMIC which can offset these). Your model is either not learning or not generating/detecting the special tokens which end a visit/end a record. Those two tokens are outside the indices of the code variables for the rest of the baselines (because those only have 2 non-pad special tokens in total and GPT uses those first two slots for start record and start visit), so a config mismatch would cause this. Otherwise the next sanity check would be that examining a patient record that gets out of the get_batch is returning those tokens as expected and the GPT model's CrossEntropyLoss is not ignoring those indices.

**Comment by jameszhou-gl** [comment]
Thanks. I ensure the gpt code is using the gpt/config.py and the code_vocab_size and total_vocab_size are correct for my mimic-iii dataset, otherwise there'll be an error raised. 
If you have time, can you help run the gpt baseline again, to exclude any potential errors from my own side? Appreciate for your assistance!

**Comment by btheodorou99** [comment]
Hello,

Upon running it myself, I am not able to reproduce this issue. Here are the steps I used:

- Cloned the repository to get the most updated code
- Updated the path to mimic and ran generateDataset.py
- Updated the config.py and baselines/gpt/config.py vocabulary sizes based on the output of that script
- Changed my working directory to baselines/gpt
- Ran train_gpt.py and test_gpt.py
- Made two minor fixes to evaluateDataset.py (which were just path errors to the gpt dataset and have been fixed now)
- Commented out everything but the gpt datasets and ran evaluateDataset
- Got expected results (printed below)

{'Record Length Mean': 1.2698692303099062, 'Record Length Standard Deviation': 0.9212053062129005, 'Visit Length Mean': 11.048574048385959, 'Visit Length Standard Deviation': 6.464637790931964}
{'Record Length Mean': 1.2620580222149038, 'Record Length Standard Deviation': 0.7539246164055974, 'Visit Length Mean': 9.588603691227405, 'Visit Length Standard Deviation': 5.780447082044173}
Per Record Code Probabilities: 0.9544123093965708
Per Visit Code Probabilities: 0.9458794861969781
Per Record Bigram Probabilities: 0.8648086529587432
Per Visit Bigram Probabilities: 0.8490177848770987
Per Record Sequential Visit Bigram Probabilities: 0.8166541487235898
Per Visit Sequential Visit Bigram Probabilities: 0.7697039428573995

**Comment by jameszhou-gl** [comment]
Thanks! I reproduce the similar results for baselines/gpt. The reason is that I wrongly set the total_vocab_size.
